mathics/core/expression.py

```diff
@@ -776,25 +788,43 @@ def evaluate(
                 "  " * evaluation.recursion_depth + "Evaluating: %s" % expr
             )
         try:
+            # Evaluation loop:
             while reevaluate:
```

Comment:
What is becoming clear is that we are doing a series of _applications_ rather than _evaluations_. 

The word "application" in functional languages is a general term that is used with the form of an expression is modified such as happens in term rewriting. And "application" is also used when you have a list and you read pick out the first element of the list and use that as a function and the remaining leaves as the function parameters. 

See for example: https://stackoverflow.com/a/8869553/546218

Reply:
In WL, Apply takes wo arguments : a head an an expression, and the result is a new expression with the same leaves of the original, but changing the head to be tee first argument.


In mathics, apply means to apply a replacement rule. Evaluate, on the other hand, is a process in which rules associated to symbols involved in a certain expression are applied iteratively until a fixed point is reached.